PythonVisualizations/Trees.py

```diff
@@ -1,538 +1,405 @@
-        self.__delete(newParent, successor, direction)
-
-    # ANIMATION METHODS
-    def speed(self, sleepTime):
-        return (sleepTime * (scaleDefault + 50)) / (scale.get() + 50)
-
-    ######For testing purposes
-    def printTree(self):
-        self.pTree(self.__root, "ROOT:  ", "")
-        print()
-
-    def pTree(self, n, kind, indent):
-        print("\n" + indent + kind, end="")
-        if n:
-            print(n, end="")
-            if n.leftChild:
-                self.pTree(n.leftChild, "LEFT:  ", indent + "    ")
-            if n.rightChild:
-                self.pTree(n.rightChild, "RIGHT:  ", indent + "    ")
-
-def clickInsert():
-    entered_text = textBox.get()
-    if entered_text:
-        val = int(entered_text)
-        if val < 100:
-            tree.insertElem(int(entered_text))
-        else:
-            outputText.set("Input value must be an integer from 0 to 99.")
-    textBox.delete(0, END)
-
-def clickFind():
-    entered_text = textBox.get()
-    if entered_text:
-        val = int(entered_text)
-        if val < 100:
-            found = tree.find(int(entered_text))
-            outputText.set("Found!" if found else "Value not found")
+            successor = self.getLeftChild(successor)
+            self.wait(0.2)
+            self.unHighlightLeftLine(newParent)
+            self.highlightLeftLine(successor)
+
+        self.wait(0.2)
+        self.unHighlightLeftLine(successor)
+        self.highlightCircle(successor)
+        self.wait(0.2)
+        self.unHighlightCircle(successor)
+
+        # delete the node that is to be deleted
+        self.removeNodeInternal(node)
+
+        # copy the items associated with successor
+        # 1. generate a new tag
+        newTag = self.generateTag()
+        # 2. copy the circle and the value
+        circle = self.copyCanvasItem(successor.drawable.display_shape)
+        circle_text = self.copyCanvasItem(successor.drawable.display_val)
+        # 3. add the appropiate tags
+        self.canvas.itemconfig(circle, tags=(newTag))
+        self.canvas.itemconfig(circle_text, tags=(newTag))
+        # 4. create a drawable
+        successorCopyDrawable = drawable(successor.getKey(), self.canvas.itemconfigure(circle, 'fill'), *(circle, circle_text))
+
+        # 5. create the lines
+        leftLine = self.copyCanvasItem(self.getLeftLine(successor))
+        rightLine = self.copyCanvasItem(self.getRightLine(successor))
+        # 6. add the appropiate tags
+        self.canvas.itemconfig(leftLine, tags=(newTag, "left line"))
+        self.canvas.itemconfig(rightLine, tags=(newTag, "right line"))
+        # 7. create the node object
+        successorCopy = Node(successorCopyDrawable, successor.coords, newTag)
+
+        # insert the successor internally
+        if direction == Child.RIGHT: self.setRightChild(parent, successorCopy)
+        else: self.setLeftChild(parent, successorCopy)
+
+        self.moveNodesToPos([successorCopy])
```

Comment:
For the animation, I'd rather show just the successor value being copied from one node to another, leaving the tree links unchanged.  Then the successor's right child and its whole subtree is moved together into position where the subtree node was.  Line 324 moves the successor with its child links up into a blank spot in the tree.  The algorithm in the textbook code, only copies the node value from one `Node` to another.

Reply:
Changes implemented in 9937d21